feat(nx): update to NgRx 8 and add schematics support for creators and entities

brandonroberts opened the pull request:
- Bumps NgRx dependencies to latest version
- Bumps jasmine-marbles dependency to `0.6.0`
- Adds a devDependency on `@ngrx/schematics` when setting up NgRx
- Adds support for NgRx creator functions in schematics and DataPersistence service
- Uses `@ngrx/entity` by default for managing collections
- Also porting setup changes upstream for less maintenance
- Updated NgRx guide and API docs

brandonroberts commented:
Still working on adding a migration to `@nrwl/angular` for breaking changes in NgRx 8. Hoping to use existing upstream migrations

wesleygrimes commented:
I'll take sometime and review this today.

wesleygrimes commented on docs/angular/guides/misc-ngrx.md:
```diff
@@ -1,484 +1,119 @@
-- `effects`,
-- `selectors`, and
-- `facade` (optional)
+The **@nrwl/angular** package has an `ngrx` schematic to generate files that implement best practices when using NgRx for state management. This schematic generates source files, along with enhancements the generated NgRx for data persistence strategies, and simplified testing.
```
```suggestion
The **@nrwl/angular** package has an `ngrx` schematic to generate files that implement best practices when using NgRx for state management. This schematic generates source files that include enhancements to NgRx for data persistence strategies, and simplified testing.
```

wesleygrimes commented on docs/angular/guides/misc-ngrx.md:
```diff
@@ -1,484 +1,119 @@
 
-- Root level NgRx configuration
-- Feature level NgRx configuration
+To get started using with NgRx in an Angular application, you set up the root level store. As your application grows, you add feature level states, ensuring that your code follows a common pattern each time.
```
```suggestion
To get started with NgRx in an Angular application, you set up the root level store. As your application grows, you add feature level states, ensuring that your code follows a common pattern each time.
```

wesleygrimes commented on docs/angular/guides/misc-ngrx.md:
```diff
@@ -1,484 +1,119 @@
-<br/>
+The following files are created, or updated:
 
-Better yet, let's generate a `Comments` feature set within a `state` library and register it with the `comments-state.module.ts` file in the same `comments/state` folder.
```
Looks like we decided to remove most of the content? Was it too wordy? I can see where value is added so that the reader can see exactly the type of content being output when these files are generated.

brandonroberts replied:
I think its too wordy to have all this output here. If there was a better way to show/hide file contents I think it would be better.

wesleygrimes commented on docs/angular/guides/misc-ngrx.md:
````diff
@@ -1,484 +1,119 @@
 
 ```bash
-ng g @nrwl/angular:ngrx Comments --root --module=apps/myapp/src/app/app.module.ts
+ng g @nrwl/angular:ngrx products --module=libs/products/src/lib/products.module.ts --directory +state/products --syntax=creators --defaults
````
Below in the API docs for the schematics options it shows "Products" in title case, but here it shows lowercase. Which is correct?

brandonroberts replied:
I've changed the references to lowercase "products" for clarity. It works either way

wesleygrimes commented on docs/api-angular/schematics/ngrx.md:
```diff
@@ -17,64 +17,88 @@ Default: `+state`
 Type: `string`
 
-Name of the NgRx feature (e.g., Products, Users, etc.). Recommended to use plural form for name.
+Name of the NgRx feature state, such as "Products" or "Users"). Recommended to use the plural form of the name.
```
Should the feature state name be title case or all lower case, or does it even matter?

wesleygrimes commented on packages/angular/src/schematics/ngrx/ngrx.spec.ts:
```diff
@@ -207,10 +230,38 @@ describe('ngrx', () => {
-      '!environment.production ? [storeFreeze] : []'
+    expect(appModule).not.toContain('!environment.production ? [] : []');
+
+    [
```
Should we also test for the existence of `state.models.ts` if the `creators` option is chosen?

wesleygrimes replied:
@brandonroberts What are your thoughts about adding a bit to test for `state.models.ts` as mentioned above?

wesleygrimes commented on packages/angular/src/schematics/ngrx/schema.json:
```diff
@@ -1,59 +1,75 @@
     "name": {
       "type": "string",
-      "description": "Name of the NgRx feature (e.g., Products, Users, etc.). Recommended to use plural form for name.",
+      "description": "Name of the NgRx feature state, such as \"Products\" or \"Users\"). Recommended to use the plural form of the name.",
```
As previously mentioned, should `Products` be title case or all lower case or does it even matter as the schematic will convert to lower?

wesleygrimes commented on packages/angular/src/utils/versions.ts:
```diff
@@ -2,7 +2,7 @@ export const nxVersion = '*';
 export const angularJsVersion = '1.6.6';
-export const ngrxVersion = '7.2.0';
+export const ngrxVersion = '8.1.0';
 export const ngrxStoreFreezeVersion = '0.2.4';
```
```suggestion
```

I believe we can remove this now as it is no longer used anywhere.

FrozenPandaz replied:
Yes, it can be removed :)

FrozenPandaz commented on packages/angular/src/schematics/ngrx/rules/add-imports-to-module.ts:
```diff
@@ -25,105 +33,136 @@ export function addImportsToModule(context: RequestContext): Rule {
+     * Delegate the effects setup to NgRx schematics
+     */
+
+    const setupRootEffects = externalSchematic('@ngrx/schematics', 'effect', {
```
Setting up root Effects seems to cause `app.module.ts` to become malformed:

![image](https://user-images.githubusercontent.com/8104246/61177091-ce86d400-a59a-11e9-9371-66c54b02260e.png)

brandonroberts replied:
The AppFacade thing is weird. I'll look into it
The `.forRoot()` and `.forFeature()` is intentional. This is to satisfy the existing use case of registering the root with a feature. In general, we want to encourage developers to keep the root object empty, and use `forFeature()` to register additional feature states.

brandonroberts commented:
@FrozenPandaz thanks for review. Is the installation of the packages not guaranteed if we modify the `package.json` and add an install task before the schematics are run?

brandonroberts commented:
I reverted the delegation to `@ngrx/schematics`, and updated the imports to follow the same general practices for setup. I also added the `--routing` option to the test to make sure it continues functioning during future changes.

FrozenPandaz commented:
Migration is broken for unrelated reasons but this looks good.

FrozenPandaz commented on packages/angular/src/migrations/update-8-3-0/upgrade-ngrx-8-0.ts:
```diff
@@ -0,0 +1,24 @@
+import { addUpdateTask, readJsonInTree, formatFiles } from '@nrwl/workspace';
+import { ngrxVersion } from '../../utils/versions';
+
+const runNgrxUpdate = addUpdateTask('@ngrx/store', ngrxVersion);
```
Let's use a hard coded version here in case people are migrating to not the latest version.

wesleygrimes commented:
Just had one outstanding question about adding to `ngrx.spec.ts` for `state.models.ts` if `creators` if chosen. Other than that I am good.

brandonroberts commented:
I have a test for the models file here https://github.com/nrwl/nx/pull/1550/files#diff-9cf1f28d0209d3021f9d79dfca3b67f2R626

wesleygrimes commented:
Perfect. I’m good with this.
